FreeCategory.agda

```diff
@@ -0,0 +1,110 @@
+module FreeCategory where
+
+open import Relation.Binary
+open import Relation.Binary.PropositionalEquality
+import Relation.Binary.SetoidReasoning as ≈-Reasoning
+
+
+module Carrier (A : Set) (C : A → A → Set) where
+
+  data ExpCategory : A → A → Set where
+    Embed : ∀ {a a′}
+          → C a a′
+          → ExpCategory a a′
+    Id    : ∀ {a}
+          → ExpCategory a a
+    _⟫_   : ∀ {a a′ a′′}
+          → ExpCategory a a′
+          → ExpCategory a′ a′′
+          → ExpCategory a a′′
+
+  data EquivCategory : ∀ {a a′}
+                     → ExpCategory a a′
+                     → ExpCategory a a′
+                     → Set where
+    equiv-refl : ∀ {a a′}
+               → {e : ExpCategory a a′}
+               → EquivCategory e e
+    equiv-sym : ∀ {a a′}
+              → {e₁ e₂ : ExpCategory a a′}
+              → EquivCategory e₁ e₂
+              → EquivCategory e₂ e₁
+    equiv-trans : ∀ {a a′}
+                → {e₁ e₂ e₃ : ExpCategory a a′}
+                → EquivCategory e₁ e₂
+                → EquivCategory e₂ e₃
+                → EquivCategory e₁ e₃
+    equiv-cong : ∀ {a a′ a′′}
+               → {e₁  e₂  : ExpCategory a a′}
+               → {e₁′ e₂′ : ExpCategory a′ a′′}
+               → EquivCategory e₁  e₂
+               → EquivCategory e₁′ e₂′
+               → EquivCategory (e₁ ⟫ e₁′) (e₂ ⟫ e₂′)
+    left-id-id : ∀ {a a′}
+               → {e : ExpCategory a a′}
+               → EquivCategory (Id ⟫ e) e
+    right-id-id : ∀ {a a′}
+                → {e : ExpCategory a a′}
+                → EquivCategory (e ⟫ Id) e
+    assoc-⟫ : ∀ {a a′ a′′ a′′′}
+            → (e   : ExpCategory a a′)
+            → (e′  : ExpCategory a′ a′′)
+            → (e′′ : ExpCategory a′′ a′′′)
+            → EquivCategory ((e ⟫ e′) ⟫ e′′) (e ⟫ (e′ ⟫ e′′))
+
+  equiv-setoid : {a a′ : A} → Setoid _ _
+  equiv-setoid {a} {a′} = record
+    { Carrier = ExpCategory a a′
+    ; _≈_ = EquivCategory
+    ; isEquivalence = record
+      { refl  = equiv-refl
+      ; sym   = equiv-sym
+      ; trans = equiv-trans
+      }
+    }
+
```

Comment:
Here's some syntactic sugar I just learned from [https://github.com/copumpkin/categories](this) repo. With it, we can write

    ≈⟨ prf1 ⟩⟫⟨ prf2 ⟩

instead of

    ≈⟨ equiv-cong prf1 prf2 ⟩